xbmc/windowing/gbm/GBMUtils.cpp

```diff
@@ -0,0 +1,555 @@
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool CGBMUtils::GetEncoder()
+{
+  for(auto i = 0; i < m_drm_resources->count_encoders; i++)
+  {
+    m_drm_encoder = drmModeGetEncoder(m_drm->fd, m_drm_resources->encoders[i]);
+    if(m_drm_encoder->encoder_id == m_drm_connector->encoder_id)
+    {
+      CLog::Log(LOGDEBUG, "CGBMUtils::%s - found encoder: %d", __FUNCTION__,
+                                                               m_drm_encoder->encoder_type);
+      break;
+    }
+    drmModeFreeEncoder(m_drm_encoder);
+    m_drm_encoder = nullptr;
+  }
+
+  if(!m_drm_encoder)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool CGBMUtils::GetPreferredMode()
+{
+  for(auto i = 0, area = 0; i < m_drm_connector->count_modes; i++)
+  {
+    drmModeModeInfo *current_mode = &m_drm_connector->modes[i];
+
+    if(current_mode->type & DRM_MODE_TYPE_PREFERRED)
+    {
+      m_drm->mode = current_mode;
+      CLog::Log(LOGDEBUG,
+                "CGBMUtils::%s - found preferred mode: %dx%d%s @ %d Hz",
+                __FUNCTION__,
+                m_drm->mode->hdisplay,
+                m_drm->mode->vdisplay,
+                m_drm->mode->flags & DRM_MODE_FLAG_INTERLACE ? "i" : "",
+                m_drm->mode->vrefresh);
+      break;
+    }
+
+    auto current_area = current_mode->hdisplay * current_mode->vdisplay;
+    if (current_area > area)
+    {
+      m_drm->mode = current_mode;
+      area = current_area;
+    }
+  }
+
+  if(!m_drm->mode)
+  {
+    CLog::Log(LOGDEBUG, "CGBMUtils::%s - failed to find preferred mode", __FUNCTION__);
+    return false;
+  }
+
+  return true;
+}
+
+drm * CGBMUtils::InitDrm()
+{
+  const char *device = "/dev/dri/card0";
+
+  m_drm->fd = open(device, O_RDWR);
+
+  if(m_drm->fd < 0)
+  {
+    return nullptr;
+  }
+
+  if(!GetResources())
+  {
+    return nullptr;
+  }
+
+  if(!GetConnector())
+  {
+    return nullptr;
+  }
+
+  if(!GetEncoder())
+  {
+    return nullptr;
+  }
+  else
+  {
+    m_drm->crtc_id = m_drm_encoder->crtc_id;
+  }
+
+  if(!GetPreferredMode())
+  {
+    return nullptr;
+  }
+
+  for(auto i = 0; i < m_drm_resources->count_crtcs; i++)
+  {
+    if(m_drm_resources->crtcs[i] == m_drm->crtc_id)
+    {
+      m_drm->crtc_index = i;
+      break;
+    }
+  }
+
+  drmModeFreeResources(m_drm_resources);
+
+  drmSetMaster(m_drm->fd);
+
+  m_drm->gbm = new gbm;
+  m_drm->gbm->dev = gbm_create_device(m_drm->fd);
+  m_drm->gbm->surface = nullptr;
+
+  m_drm_fds.fd = m_drm->fd;
+  m_drm_fds.events = POLLIN;
+
+  m_drm_evctx.version = DRM_EVENT_CONTEXT_VERSION;
+  m_drm_evctx.page_flip_handler = PageFlipHandler;
+
+  m_drm->connector_id = m_drm_connector->connector_id;
+  m_orig_crtc = drmModeGetCrtc(m_drm->fd, m_drm->crtc_id);
+
+  return m_drm;
+}
+
+bool CGBMUtils::RestoreOriginalMode()
+{
+  if(!m_orig_crtc)
+  {
+    return false;
+  }
+
+  auto ret = drmModeSetCrtc(m_drm->fd,
+                            m_orig_crtc->crtc_id,
+                            m_orig_crtc->buffer_id,
```

Comment:
The original framebuffer might be long gone by the time you get here: failure should not be fatal.